OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,274 @@
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+/** @file
+* This file defines the  DataTable class, which is used by OpenSim to provide
+* an in memory container for data access and manipulation.
+*/
+#include "SimTKmath.h"
+#include "XMLDocument.h"
+#include "FileAdapter.h"
+#include "Array.h"
+#include <memory>
+#include <typeindex>
+
+
+namespace OpenSim {
+
+//=============================================================================
+/**
+ * AbstractDataTable defines a simple interface for in memory access to
+ * numerical data as a table. A DataTable is independent of the data source 
+ * used to populate it. A concrete DataTable provides fast access to data
+ * elements by row and/or column indices as well as columns by label.
+ *
+ * @author Ajay Seth
+ */
+class AbstractDataTable {
+public:
+    virtual ~AbstractDataTable() = default;
+
+    virtual AbstractDataTable* clone() const = 0;
+
+    /** Read the meta data associated with this DataTable as an XMLDocument */
+    virtual const XMLDocument& getMetaData() const = 0;
+    /** Update the meta data associated with this DataTable as an XMLDocument */
+    virtual XMLDocument& updMetaData() = 0;
+
+    /** Get the number of rows in the DataTable */
+    virtual size_t  getNumRows() const = 0;
+    /** Get the number of columns in the DataTable */
+    virtual size_t  getNumCols() const = 0;
+
+    /** Determine if the DataTable has a column associated with a 
+        column label (string) */
+    virtual bool hasColumn(const std::string& label) const = 0;
+    /** For a known column label return its column index. */
+    virtual size_t getColumnIndex(const std::string& label) const = 0;
+
+    /** Read-only column labels associated with the columns of the DataTable */
+    virtual const Array<std::string>& getColumnLabels() const = 0;
+    /** Update column labels associated with the columns of the DataTable */
+    virtual Array<std::string>& updColumnLabels() = 0;
+
+    /* TODO: Read in the units associated with the columns of the DataTable */
+    /* virtual const Array<Units>& getColumnUnits() const = 0;
+    */
+
+    /** Get the underlying type of data held in the DataTable. Can be any 
+        SimTK numerical type: Real, Vec<M>, Mat<M, N>, SpatialVec, ... */
+    virtual const std::type_index& getDataTypeInfo() const = 0;
```

Comment:
I think it would be better just to return a reference to an `std::type_info` object here. Because that is a static object, there would be no need for the derived class to store it. And anyone who needs a `std::type_index` for it can easily get one with `std::type_index(getDataTypeInfo())`.